File: frontend/lib/features/students/domain/services/bulk_teacher_action_service.dart
import '../../../lessons/domain/entities/entities.dart';
import '../../../lessons/domain/repositories/lesson_repository.dart';
import '../../../schedule/domain/entities/request_event.dart';
import '../../../schedule/domain/entities/unified_lesson_request.dart';
import '../../../schedule/domain/repositories/unified_lesson_request_repository.dart';
import '../../../subscription/domain/entities/subscription.dart';
import '../../../subscription/domain/repositories/subscription_repository.dart';
import '../entities/bulk_cancel_result.dart';

/// §7.119 v2 Bulk Teacher Actions — 선택 모드에서 선생님이 여러 학생에게 동시 실행하는 운영 작업.
///
/// 지원 유스케이스:
/// - B1 휴강 공지: 특정 날짜의 레슨을 일괄 취소 + RequestEvent
/// - B2 일괄 메시지: 학생들에게 공지 RequestEvent 브로드캐스트
///
/// v2 변경 (2026-05-07):
/// - 회차(sessionNumber) 매핑 + 수강권 챗 이벤트(RequestEvent) 생성
/// - 수강권 미발급 학생 자동 제외
/// - 일괄 메시지 수신 조건 필터 (활성 수강권만 / 전체)
class BulkTeacherActionService {
  final LessonRepository _lessonRepository;
  final UnifiedLessonRequestRepository _requestRepository;
  final SubscriptionRepository _subscriptionRepository;

  BulkTeacherActionService({
    required LessonRepository lessonRepository,
    required UnifiedLessonRequestRepository requestRepository,
    required SubscriptionRepository subscriptionRepository,
  }) : _lessonRepository = lessonRepository,
       _requestRepository = requestRepository,
       _subscriptionRepository = subscriptionRepository;

  /// 대상 날짜에 예정된 레슨을 일괄 취소하고 각 학생의 수강권 챗에 취소 이벤트를 남긴다.
  ///
  /// - 대상 날짜에 scheduled 상태 레슨이 없는 학생은 [BulkCancelResult.skippedStudentIds] 에 포함
  /// - 이미 취소 상태인 레슨은 건드리지 않는다 (중복 취소 방지)
  /// - 상태는 [LessonStatus.cancelledByTeacher] 로 명시 설정 → 차감 없음(`isDeducted=false`)
  ///   + 학생 측 reschedule 요청 가능(`allowsReschedule=true`) 보장. 일반
  ///   `cancelLesson()` 는 generic [LessonStatus.cancelled] 로 떨어져 reschedule 경로가
  ///   막히기 때문에 사용하지 않는다.
  /// 대상 날짜에 **수업이 있는 학생만** 일괄 취소.
  ///
  /// 필터링 순서:
  /// 1. targetDate에 예정된 레슨 조회 (`getLessonsByDate`)
  /// 2. `scheduled` / `reschedulePending` 상태만 (`_isCancellable`)
  /// 3. 선택된 studentIds에 포함된 학생만
  /// 4. 활성 수강권 없는 학생은 skipped (수강권 미발급 제외)
  ///
  /// → **수업일이 아닌 학생**은 자동으로 skipped에 포함.
  Future<BulkCancelResult> cancelLessonsOnDate({
    required String teacherId,
    required List<String> studentIds,
    required DateTime targetDate,
    String? reason,
  }) async {
    final lessonsOnDate = await _lessonRepository.getLessonsByDate(targetDate);

    // 활성 수강권 보유 학생만 필터
    final activeStudents = await _filterActiveSubscriptionStudents(studentIds);
    final activeStudentSet = activeStudents.toSet();

    final lessonsByStudent = <String, Lesson>{
      for (final lesson in lessonsOnDate)
        if (_isCancellable(lesson) &&
            activeStudentSet.contains(lesson.studentId))
          lesson.studentId: lesson,
    };

    final skipped = <String>[];
    for (final studentId in studentIds) {
      if (!lessonsByStudent.containsKey(studentId)) {
        skipped.add(studentId);
      }
    }

    var cancelledCount = 0;
    var notifiedCount = 0;
    for (final entry in lessonsByStudent.entries) {
      final lesson = entry.value;
      await _lessonRepository.updateLesson(
        lesson.copyWith(
          status: LessonStatus.cancelledByTeacher,
          updatedAt: DateTime.now(),
        ),
      );
      cancelledCount++;

      // v2: 수강권 챗에 RequestEvent 생성 (회차 매핑)
      // #1212 — 학생 통지는 이 챗 이벤트(BE) 가 SSOT.
      // flutter_local_notifications 는 액터(교사) 기기 전용이라 상대 통지로 쓸 수
      // 없어(기존엔 교사 기기에 오발) FE 로컬 알림 호출을 제거함.
      await _createCancelEvent(
        teacherId: teacherId,
        lesson: lesson,
        reason: reason,
      );
      notifiedCount++;
    }

    return BulkCancelResult(
      cancelledLessonCount: cancelledCount,
      notifiedStudentCount: notifiedCount,
      skippedStudentIds: skipped,
    );
  }

  /// 선생님 공지 메시지를 학생들의 수강권 챗에 teacherAnnouncement 이벤트로 발송.
  ///
  /// [activeOnly] = true: 활성 수강권 보유 학생만 (기본)
  /// [activeOnly] = false: 선택된 전체 학생
  ///
  /// 반환: 공지를 전달한 학생 수.
  Future<int> broadcastMessage({
    required String teacherId,
    required List<String> studentIds,
    required String title,
    required String body,
    bool activeOnly = true,
  }) async {
    if (studentIds.isEmpty) return 0;

    // 수신 대상 필터링
    final targetIds = activeOnly
        ? await _filterActiveSubscriptionStudents(studentIds)
        : studentIds;

    var sent = 0;
    for (final studentId in targetIds) {
      // #1212 — 학생 통지는 챗 RequestEvent(BE) 가 SSOT.
      // flutter_local_notifications 는 액터(교사) 기기 전용이라 상대 통지로 쓸 수
      // 없어(기존엔 교사 기기에 오발) FE 로컬 알림 호출을 제거함.
      await _createAnnouncementEvent(
        teacherId: teacherId,
        studentId: studentId,
        title: title,
        body: body,
      );

      sent++;
    }
    return sent;
  }

  /// 대상 날짜에서 취소 대상이 될 레슨 목록을 mutation 없이 반환 (확인 다이얼로그용).
  Future<List<Lesson>> previewAffectedLessons({
    required List<String> studentIds,
    required DateTime targetDate,
  }) async {
    final lessons = await _lessonRepository.getLessonsByDate(targetDate);
    return lessons
        .where((l) => _isCancellable(l) && studentIds.contains(l.studentId))
        .toList();
  }

  bool _isCancellable(Lesson lesson) {
    switch (lesson.status) {
      case LessonStatus.scheduled:
      case LessonStatus.reschedulePending:
        return true;
      default:
        return false;
    }
  }

  /// v2: 수강권 챗에 lessonCancelledByTeacher 이벤트 생성.
  Future<void> _createCancelEvent({
    required String teacherId,
    required Lesson lesson,
    String? reason,
  }) async {
    final subscriptionId = lesson.subscriptionId;
    if (subscriptionId == null) return;

    final sub = await _subscriptionForLesson(lesson);
    final requestId = await _requestIdForStudentTeacher(
      studentId: lesson.studentId,
      teacherId: teacherId,
    );
    if (requestId == null) return;

    final event = RequestEvent(
      id: 'cancel_event_${DateTime.now().millisecondsSinceEpoch}_${lesson.id}',
      requestId: requestId,
      actorType: ProposerRole.teacher,
      actorId: teacherId,
      eventType: RequestEventType.lessonCancelledByTeacher,
      subscriptionId: subscriptionId,
      sessionNumber: _sessionNumberFor(sub),
      message: reason,
      changeCreditUsed: 0,
      changeCreditRemainingAfter: sub?.remainingReschedule,
      keepsSessionNumber: true,
      createdAt: DateTime.now(),
    );
    await _requestRepository.addEvent(event);
  }

  /// v2: 수강권 챗에 teacherAnnouncement 이벤트 생성.
  Future<void> _createAnnouncementEvent({
    required String teacherId,
    required String studentId,
    required String title,
    required String body,
  }) async {
    final requestId = await _requestIdForStudentTeacher(
      studentId: studentId,
      teacherId: teacherId,
    );
    if (requestId == null) return;

    final subs = await _subscriptionRepository.getByStudentId(studentId);
    final activeSub = subs.where(_isActiveSubscription).firstOrNull;

    final event = RequestEvent(
      id: 'announce_${DateTime.now().millisecondsSinceEpoch}_$studentId',
      requestId: requestId,
      actorType: ProposerRole.teacher,
      actorId: teacherId,
      eventType: RequestEventType.teacherAnnouncement,
      subscriptionId: activeSub?.id,
      message: '$title\n$body',
      createdAt: DateTime.now(),
    );
    await _requestRepository.addEvent(event);
  }

  /// v2: 활성 수강권 보유 학생만 필터.
  Future<List<String>> _filterActiveSubscriptionStudents(
    List<String> studentIds,
  ) async {
    final result = <String>[];
    for (final studentId in studentIds) {
      final subs = await _subscriptionRepository.getByStudentId(studentId);
      if (subs.any(_isActiveSubscription)) {
        result.add(studentId);
      }
    }
    return result;
  }

  bool _isActiveSubscription(Subscription sub) {
    return sub.status == SubscriptionStatus.active ||
        sub.status == SubscriptionStatus.expiringSoon;
  }

  Future<Subscription?> _subscriptionForLesson(Lesson lesson) async {
    final subscriptionId = lesson.subscriptionId;
    if (subscriptionId == null) return null;

    final subs = await _subscriptionRepository.getByStudentId(lesson.studentId);
    return subs.where((s) => s.id == subscriptionId).firstOrNull;
  }

  int? _sessionNumberFor(Subscription? subscription) {
    if (subscription == null) return null;
    return subscription.usedLessons + 1;
  }

  Future<String?> _requestIdForStudentTeacher({
    required String studentId,
    required String teacherId,
  }) async {
    final requests = await _requestRepository.getByStudentId(studentId);
    final matching =
        requests.where((request) => request.teacherId == teacherId).toList()
          ..sort((a, b) => b.createdAt.compareTo(a.createdAt));
    return matching.firstOrNull?.id;
  }
}

File: backend/app/services/bulk_teacher_action_service.py
"""Bulk teacher action orchestration."""

from __future__ import annotations

from typing import Any

from fastapi import HTTPException, status
from sqlalchemy import and_, or_, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.lesson import LessonStatus
from app.models.notification import NotificationPriority
from app.models.request_event import RequestEventType
from app.models.subscription import SubscriptionStatus, SubscriptionType
from app.schemas.lesson import BulkCancelLessonResponse, BulkLessonEventCreated
from app.schemas.notification import BroadcastNotificationResponse
from app.services.notification_service import NotificationService
from app.services.teacher_id_resolver import resolve_teacher_id

ACTIVE_SUBSCRIPTION_STATUSES = {
    SubscriptionStatus.active,
    SubscriptionStatus.expiringSoon,
}


class BulkTeacherActionService:
    """Handle teacher bulk cancel and broadcast actions."""

    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def bulk_cancel_lessons(
        self,
        *,
        teacher_id: str,
        student_ids: list[str],
        target_date: Any,
        reason: str | None,
        notification_title: str,
        current_user: Any,
    ) -> BulkCancelLessonResponse:
        """Cancel matching lessons and create subscription chat events."""
        return await self._bulk_cancel_lessons(
            teacher_id=teacher_id,
            student_ids=student_ids,
            target_date=target_date,
            reason=reason,
            notification_title=notification_title,
            current_user=current_user,
            preview=False,
        )

    async def preview_bulk_cancel_lessons(
        self,
        *,
        teacher_id: str,
        student_ids: list[str],
        target_date: Any,
        reason: str | None,
        notification_title: str,
        current_user: Any,
    ) -> BulkCancelLessonResponse:
        """Preview matching lesson cancellations without side effects."""
        return await self._bulk_cancel_lessons(
            teacher_id=teacher_id,
            student_ids=student_ids,
            target_date=target_date,
            reason=reason,
            notification_title=notification_title,
            current_user=current_user,
            preview=True,
        )

    async def _bulk_cancel_lessons(
        self,
        *,
        teacher_id: str,
        student_ids: list[str],
        target_date: Any,
        reason: str | None,
        notification_title: str,
        current_user: Any,
        preview: bool,
    ) -> BulkCancelLessonResponse:
        from app.models.lesson import Lesson
        from app.models.request_event import RequestEvent

        resolved_teacher_id = await self._authorize_teacher(teacher_id, current_user)
        students = await self._students_by_id(resolved_teacher_id, student_ids)
        subscriptions = await self._active_subscriptions_by_student(resolved_teacher_id, student_ids)

        result = await self.db.scalars(
            select(Lesson).where(
                Lesson.teacher_id == resolved_teacher_id,
                Lesson.student_id.in_(student_ids),
                Lesson.date == target_date,
                Lesson.status.in_([LessonStatus.scheduled, LessonStatus.reschedulePending]),
            )
        )
        lessons = result.all()
        lessons_by_student: dict[str, list[Any]] = {}
        for lesson in lessons:
            lessons_by_student.setdefault(lesson.student_id, []).append(lesson)

        events_created: list[BulkLessonEventCreated] = []
        notified_students: set[str] = set()
        skipped_student_ids: list[str] = []
        notification_service = NotificationService(self.db)

        for student_id in student_ids:
            student = students.get(student_id)
            subscription = subscriptions.get(student_id)
            student_lessons = lessons_by_student.get(student_id, [])
            if student is None or subscription is None or not student_lessons:
                skipped_student_ids.append(student_id)
                continue

            for lesson in student_lessons:
                session_number = await self._session_number_for_lesson(lesson)
                if not preview:
                    lesson.status = LessonStatus.cancelledByTeacher
                    event = RequestEvent(
                        request_id=subscription.id,
                        actor_type="teacher",
                        actor_id=current_user.id,
                        event_type=RequestEventType.lesson_cancelled_by_teacher,
                        subscription_id=subscription.id,
                        session_number=session_number,
                        message=reason,
                        change_credit_used=0,
                        change_credit_remaining_after=(
                            subscription.total_reschedule_allowance - subscription.used_reschedule_count
                        ),
                        keeps_session_number=True,
                    )
                    self.db.add(event)
                    body = self._bulk_cancel_body(lesson, reason, session_number)
                    await notification_service.create_and_send(
                        user_id=student.user_id or student.id,
                        notification_type="lessonCancelled",
                        title=notification_title,
                        body=body,
                        priority=NotificationPriority.high,
                        data={
                            "teacherId": resolved_teacher_id,
                            "lessonId": lesson.id,
                            "subscriptionId": subscription.id,
                            "sessionNumber": session_number,
                            "source": "bulk_teacher_action",
                            "actionUrl": f"/subscriptions/{subscription.id}",
                            "actionLabel": "보강 요청",
                        },
                        action_url=f"/subscriptions/{subscription.id}",
                        action_label="보강 요청",
                    )
                events_created.append(
                    BulkLessonEventCreated(
                        student_id=student_id,
                        lesson_id=lesson.id,
                        session_number=session_number,
                        subscription_id=subscription.id,
                    )
                )
                notified_students.add(student_id)

        await self.db.flush()
        return BulkCancelLessonResponse(
            cancelled_lesson_count=len(events_created),
            notified_student_count=len(notified_students),
            skipped_student_ids=skipped_student_ids,
            events_created=events_created,
        )

    async def broadcast(
        self,
        *,
        teacher_id: str,
        student_ids: list[str],
        target_filter: str,
        title: str,
        body: str,
        current_user: Any,
    ) -> BroadcastNotificationResponse:
        """Send a teacher broadcast and create chat events for active subscriptions."""
        from app.models.request_event import RequestEvent

        resolved_teacher_id = await self._authorize_teacher(teacher_id, current_user)
        students = await self._students_by_id(resolved_teacher_id, student_ids)
        subscriptions = await self._active_subscriptions_by_student(resolved_teacher_id, student_ids)
        notification_service = NotificationService(self.db)

        sent_count = 0
        event_created_count = 0
        filtered_out_count = 0

        for student_id in student_ids:
            student = students.get(student_id)
            subscription = subscriptions.get(student_id)
            if student is None:
                filtered_out_count += 1
                continue
            if target_filter == "active_subscription" and subscription is None:
                filtered_out_count += 1
                continue

            if subscription is not None:
                self.db.add(
                    RequestEvent(
                        request_id=subscription.id,
                        actor_type="teacher",
                        actor_id=current_user.id,
                        event_type=RequestEventType.teacher_announcement,
                        subscription_id=subscription.id,
                        message=f"{title}\n{body}",
                    )
                )
                event_created_count += 1

            await notification_service.create_and_send(
                user_id=student.user_id or student.id,
                notification_type="generalAnnouncement",
                title=title,
                body=body,
                priority=NotificationPriority.normal,
                data={
                    "teacherId": resolved_teacher_id,
                    "subscriptionId": subscription.id if subscription is not None else None,
                    "source": "bulk_teacher_action",
                },
            )
            sent_count += 1

        await self.db.flush()
        return BroadcastNotificationResponse(
            sent_count=sent_count,
            event_created_count=event_created_count,
            filtered_out_count=filtered_out_count,
        )

    async def _authorize_teacher(self, teacher_id: str, current_user: Any) -> str:
        resolved_teacher_id = await resolve_teacher_id(self.db, current_user.id)
        if teacher_id not in {current_user.id, resolved_teacher_id}:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Teacher access denied")
        return resolved_teacher_id

    async def _students_by_id(self, teacher_id: str, student_ids: list[str]) -> dict[str, Any]:
        from app.models.student import Student

        result = await self.db.scalars(
            select(Student).where(
                Student.teacher_id == teacher_id,
                Student.id.in_(student_ids),
                Student.is_active.is_(True),
            )
        )
        return {student.id: student for student in result.all()}

    async def _active_subscriptions_by_student(self, teacher_id: str, student_ids: list[str]) -> dict[str, Any]:
        from app.models.lesson import ClassMembership, LessonClass
        from app.models.subscription import Subscription

        result = await self.db.scalars(
            select(Subscription)
            .join(ClassMembership, Subscription.membership_id == ClassMembership.id)
            .join(LessonClass, ClassMembership.lesson_class_id == LessonClass.id)
            .where(
                LessonClass.teacher_id == teacher_id,
                Subscription.student_id.in_(student_ids),
                or_(
                    Subscription.status.in_(ACTIVE_SUBSCRIPTION_STATUSES),
                    and_(
                        Subscription.type == SubscriptionType.trial,
                        Subscription.status.in_(ACTIVE_SUBSCRIPTION_STATUSES),
                    ),
                ),
            )
            .order_by(Subscription.created_at.desc(), Subscription.id.desc())
        )
        subscriptions_by_student: dict[str, Any] = {}
        for subscription in result.all():
            subscriptions_by_student.setdefault(subscription.student_id, subscription)
        return subscriptions_by_student

    async def _session_number_for_lesson(self, lesson: Any) -> int | None:
        from app.models.lesson import Lesson

        if lesson.subscription_id is None:
            return None

        result = await self.db.scalars(
            select(Lesson)
            .where(Lesson.subscription_id == lesson.subscription_id)
            .order_by(Lesson.date.asc(), Lesson.start_time.asc(), Lesson.id.asc())
        )
        for index, subscription_lesson in enumerate(result.all(), start=1):
            if subscription_lesson.id == lesson.id:
                return index
        return None

    def _bulk_cancel_body(self, lesson: Any, reason: str | None, session_number: int | None) -> str:
        session_text = f"{session_number}회차 " if session_number is not None else ""
        body = f"{session_text}{lesson.date.isoformat()} {lesson.start_time} 레슨이 휴강 처리되었습니다."
        if reason:
            body = f"{body}\n사유: {reason}"
        return body
